syncstorage/storage/spanner.py

```diff
@@ -0,0 +1,987 @@
+                    "collectionid": collectionid},
+            param_types={"userid": param_types.STRING,
+                         "collectionid": param_types.INT64}
+        ).one_or_none()
+        if result is None:
+            raise CollectionNotFoundError
+        return dt2ts(result[0])
+
+    @with_session
+    def get_items(self, session, user, collection, **params):
+        """Returns items from a collection."""
+        return self._find_items(session, user, collection, **params)
+
+    @with_session
+    def get_item_ids(self, session, user, collection, **params):
+        """Returns item ids from a collection."""
+        # Select only the fields we need, including those that might
+        # be used for limit/offset pagination.
+        params["fields"] = ["id", "modified", "sortindex"]
+        res = self._find_items(session, user, collection, **params)
+        res["items"] = [item["id"] for item in res["items"]]
+        return res
+
+    def _find_items(self, session, user, collection, **params):
+        """Find items matching the given search parameters."""
+        userid = user_key(user)
+        collectionid = self._get_collection_id(collection)
+        bind = {"userid": userid,
+                "collectionid": collectionid}
+        bind_types = {"userid": param_types.STRING,
+                      "collectionid": param_types.INT64}
+        ts = get_timestamp()
+        if "ttl" not in params:
+            params["ttl"] = ts
+
+        if "ids" in params:
+            for i, id in enumerate(params["ids"], 1):
+                bind["id_%d" % (i,)] = id
+                bind_types["id_%d" % (i,)] = param_types.STRING
+
+        if "sort" in params:
+            bind["sort"] = params["sort"]
+            bind_types["sort"] = param_types.STRING
+
+        # We always fetch one more item than necessary, so we can tell whether
+        # there are additional items to be fetched with next_offset.
+        limit = params.get("limit")
+        if limit is not None:
+            params["limit"] = bind["limit"] = limit + 1
+            bind_types["limit"] = param_types.INT64
+        offset = params.pop("offset", None)
+        if offset is not None:
+            self.decode_offset(params, offset)
+            bind["offset"] = params["offset"]
+            bind_types["offset"] = param_types.INT64
+            if limit is None:
+                # avoid sqlalchemy defaulting to LIMIT -1 when none provided
+                params["limit"] = bind["limit"] = INT64_MAX
+                bind_types["limit"] = param_types.INT64
+
+        # Convert timestamp types
+        for ts_var in ["newer", "newer_eq", "older", "older_eq", "ttl"]:
+            if ts_var in params:
+                bind[ts_var] = ts2dt(params[ts_var])
+                bind_types[ts_var] = param_types.TIMESTAMP
+
+        # Generate the query
+        query = FIND_ITEMS(bso, params)
+        query = str(query.compile()).replace(":", "@")
```

Comment:
As above, is there any chance of this `.replace` overzealously matching something it shouldn't? I can't think of any concrete example off the top of my head, just wondering out loud...

Reply:
There's no way for a user to insert data into our query, as we're using bind params for all variables, and the existing queries use : purely for binding (or they'd fail as well). So the only way I could see that this would cause problems is if the original queries have invalid variable bindings.